src/main/java/org/jenkinsci/plugins/workflow/log/FileLogStorage.java

```diff
@@ -0,0 +1,276 @@
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package org.jenkinsci.plugins.workflow.log;
+
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import hudson.console.AnnotatedLargeText;
+import hudson.console.ConsoleAnnotationOutputStream;
+import hudson.model.BuildListener;
+import hudson.model.StreamBuildListener;
+import hudson.model.TaskListener;
+import hudson.util.StreamTaskListener;
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.FilterOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.io.OutputStreamWriter;
+import java.io.RandomAccessFile;
+import java.io.Writer;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.logging.Logger;
+import org.apache.commons.io.input.NullReader;
+import org.jenkinsci.plugins.workflow.flow.FlowExecutionOwner;
+import org.jenkinsci.plugins.workflow.graph.FlowNode;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.Beta;
+import org.kohsuke.stapler.framework.io.ByteBuffer;
+
+/**
+ * Simple implementation of log storage in a single file that maintains a side file with an index indicating where node transitions occur.
+ * Each line in the index file is a byte offset, optionally followed by a space and then a node ID.
+ */
+@Restricted(Beta.class)
+public final class FileLogStorage implements LogStorage {
+
+    private static final Logger LOGGER = Logger.getLogger(FileLogStorage.class.getName());
+
+    private static final Map<File, FileLogStorage> openStorages = Collections.synchronizedMap(new HashMap<>());
+
+    public static synchronized LogStorage forFile(File log) {
+        return openStorages.computeIfAbsent(log, FileLogStorage::new);
+    }
+
+    private final File log;
+    private final File index;
+    @SuppressFBWarnings(value = "IS2_INCONSISTENT_SYNC", justification = "FB apparently gets confused by what the lock is, and anyway we only care about synchronizing writes")
+    private FileOutputStream os;
+    private Writer indexOs;
+    private String lastId;
+
+    private FileLogStorage(File log) {
+        this.log = log;
+        this.index = new File(log + "-index");
+    }
+
+    private synchronized void open() throws IOException {
```

Comment:
@jglick Would be great to get a comment to explain this -- especially curious if there's a specific reason why we're reading the entire file line by line in order to get the last line, rather than using a RandomAccessFile or similar and scanning in blocks from the end looking for a new line separator...

Reply:
Actually, scratch that, couldn't you just look for a space?

Reply:
This was just the simplest code to write. If there is any specific reason to suspect a performance issue, it could of course be optimized to scan the file backwards for the penultimate `\n`, then convert the byte sequence from there to EOF to UTF-8 and set `lastId` accordingly.